text/0000-explicit-tail-calls.md

```diff
@@ -0,0 +1,648 @@
+
+The goal behind this design is to allow TCE of functions that do not have exactly matching function signatures, in
+theory, this just requires that tail-called functions are callee cleanup, which is a mismatch to the default calling
+convention used by Rust. To limit the impact of this change all functions that should be TCE-able should be marked with
+an attribute.
+
+While quite noisy it is also less flexible than the chosen approach. Indeed, TCE is a property of the call and not a
+function definition, sometimes a call should be guaranteed to be TCE, and sometimes not, marking a function would
+be less flexible.
+
+### Adding a mark to `return`.
+
+The return keyword could be marked using an attribute or an extra keyword as in the example below.
+
+```rust
+fn a() {
+    // The chosen variant.
+    become b();
+
+    // Using an attribute.
+    #[become]
+    return b();
+
+    // Adding an extra keyword.
+    return become b();
+}
+```
+
+These alternatives mostly come down to personal taste (or bikeshedding) and the plain keyword `become` was chosen because of the following reasons:
+
+- It is [reserved](https://rust-lang.github.io/rfcs/0601-replace-be-with-become.html) exactly this use case.
+- It is shorter to write.
+- The behavior changes in subtle ways compared to a plain `return`. To clearly indicate this change in behavior a stronger distinction from `return` than adding a mark seems warranted.
+  - TCE as proposed in this RFC requires dropping local variables before the function call instead of after with `return`.
+  - From a type system perspective the type of the `return` expression (`!`, the never type, see [here](https://doc.rust-lang.org/std/primitive.never.html) for an example) stays the same even when adding one of the markings. This means that type-checking can not help if the marking is forgotten or added mistakenly. (Note that the argument, the function call, can still be type checked, just not the `return` expression.)
+   
+### Require Explicit Dropping of Variables
+
+(Based on this [comment](https://github.com/rust-lang/rfcs/pull/3407#issuecomment-1532841475))
+
+An alternative approach could be to refuse to compile functions that would need to run destructors before `become`.
+This would force code to not rely on implicit drops and require calls to `drop(variable)`, as in the following example:
+
+```rust
+fn f(x: String) {
+    drop(x); // necessary
+    become g();
+}
+```
+
+This approach would result in more verbose code but would also be easier to read for people not familiar with tail calls.
+Also, it is forwards compatible with implicit dropping before `become`.
+
+The reason this approach is not chosen is that the tradeoff between increased verbosity and the reduction of initial learning time seems to not be worth it.
```

Comment:
I want to add that the implementation of the diagnostic for forgotten drops would likely be *harder*, than the implementation of correct drop elaboration. (this doesn't *necessarily* need to be added to the RFC, but just though I'd comment)

Reply:
added here: https://github.com/rust-lang/rfcs/pull/3407/commits/0e4a645a87243116fa9d84bf0f77047cc5b79df7

Added, just in case this alternative wasn't unappealing enough. :sweat_smile: